psci/Main.hs

```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE DoAndIfThenElse #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE DataKinds #-}
+
+module Main (main) where
+
+import           Prelude ()
+import           Prelude.Compat
+
+import           Data.Monoid ((<>))
+import           Data.Version (showVersion)
+
+import           Control.Applicative (many)
+import           Control.Monad
+import           Control.Monad.Trans.Class
+import           Control.Monad.Trans.Except (ExceptT(..), runExceptT)
+import           Control.Monad.Trans.State.Strict (StateT, evalStateT)
+
+import qualified Language.PureScript as P
+import           Language.PureScript.Interactive
+
+import qualified Options.Applicative as Opts
+
+import qualified Paths_purescript as Paths
+
+import           System.Console.Haskeline
+import           System.Exit
+import           System.FilePath.Glob (glob)
+import           System.IO.UTF8 (readUTF8File)
+
+multiLineMode :: Opts.Parser Bool
+multiLineMode = Opts.switch $
+     Opts.long "multi-line-mode"
+  <> Opts.short 'm'
+  <> Opts.help "Run in multi-line mode (use ^D to terminate commands)"
+
+inputFile :: Opts.Parser FilePath
+inputFile = Opts.strArgument $
+     Opts.metavar "FILE"
+  <> Opts.help "Optional .purs files to load on start"
+
+inputForeignFile :: Opts.Parser FilePath
+inputForeignFile = Opts.strOption $
+     Opts.short 'f'
+  <> Opts.long "ffi"
+  <> Opts.help "The input .js file(s) providing foreign import implementations"
+
+nodeFlagsFlag :: Opts.Parser [String]
+nodeFlagsFlag = Opts.option parser $
+     Opts.long "node-opts"
+  <> Opts.metavar "NODE_OPTS"
+  <> Opts.value []
+  <> Opts.help "Flags to pass to node, separated by spaces"
+  where
+    parser = words <$> Opts.str
+
+psciOptions :: Opts.Parser PSCiOptions
+psciOptions = PSCiOptions <$> multiLineMode
+                          <*> many inputFile
+                          <*> many inputForeignFile
+                          <*> nodeFlagsFlag
+
+version :: Opts.Parser (a -> a)
+version = Opts.abortOption (Opts.InfoMsg (showVersion Paths.version)) $
+            Opts.long "version" <>
+            Opts.help "Show the version number" <>
+            Opts.hidden
+
+getOpt :: IO PSCiOptions
+getOpt = Opts.execParser opts
+    where
+      opts        = Opts.info (version <*> Opts.helper <*> psciOptions) infoModList
+      infoModList = Opts.fullDesc <> headerInfo <> footerInfo
+      headerInfo  = Opts.header   "psci - Interactive mode for PureScript"
+      footerInfo  = Opts.footer $ "psci " ++ showVersion Paths.version
+
+-- | Read a file in the 'P.Make' monad, handling errors.
+readFileMake :: FilePath -> P.Make String
+readFileMake path = P.makeIO (const (P.ErrorMessage [] $ P.CannotReadFile path)) (readUTF8File path)
```

Comment:
Perhaps we should export `readTextFile` from `Language.PureScript.Make` instead of redefining it here?